### PR TITLE
deno 1.23.1

### PR DESCRIPTION
**Pull request by branchv**
Created with `brew bump-formula-pr`.

~`resource` blocks may require updates.~

### CODE REVIEW

**Comment by chenrui333** [comment]
might need rust 1.16.x

**Comment by matthewleon** [comment]
What's stopping this from being merged? It looks like some CI task is failing non-deterministically?

**Comment by SMillerDev** [comment]
It's failing to build, that's why it's not merged.

**Comment by ry** [comment]
This appears to be because the latest V8 depends on `-Wno-unqualified-std-cast-call` a new clang flag. I'm not sure when it was introduced but we're using clang 15 to build V8.

**Comment by aslilac** [comment]
Could we run a build of 1.23 to see what the current state is?

**Comment by tylerbrostrom** [comment]
https://github.com/denoland/rusty_v8/pull/1009 appears to unblock the failing build issue

**Comment by aslilac** [comment]
> [denoland/rusty_v8#1009](https://github.com/denoland/rusty_v8/pull/1009) appears to unblock the failing build issue

Still needs to be upstreamed tho. We're planning to have it in for the 1.23.1 release tomorrow.

**Comment by lucacasonato** [comment]
1.23.1 is now released. Can you try build again with this version?

File: Formula/deno.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 class Deno < Formula
   desc "Secure runtime for JavaScript and TypeScript"
   homepage "https://deno.land/"
-  url "https://github.com/denoland/deno/releases/download/v1.22.0/deno_src.tar.gz"
-  sha256 "ffa1cd143ba1e53ab5d380c1e630a2fe73910b7be4e483b0d643d14b31b23ed4"
+  url "https://github.com/denoland/deno/releases/download/v1.23.0/deno_src.tar.gz"
+  sha256 "0432681c1c88101f23fa0d927f115413455c10ea4ba801d322a9eb52180763d1"
```
Comment (aslilac) [inline Formula/deno.rb:5]:
```suggestion
  url "https://github.com/denoland/deno/releases/download/v1.23.1/deno_src.tar.gz"
  sha256 "619169f952f1730870a7d1039bf3d65f030f520f5a7d2f50673fd1899ceeace7"
```

File: Formula/deno.rb [hunk 2]
```diff
@@ -40,7 +40,7 @@ class Deno < Formula
   # 4. Find full gn commit hash: https://gn.googlesource.com/gn.git/+/#{gn_commit}
   resource "gn" do
     url "https://gn.googlesource.com/gn.git",
-        revision: "53d92014bf94c3893886470a1c7c1289f8818db0"
+        revision: "bf4e17dc67b2a2007475415e3f9e1d1cf32f6e35"
```
Comment (aslilac) [inline Formula/deno.rb:43]: gn revision is still correct 👍

File: Formula/deno.rb (hunk 1 above)
Comment (branchvincent) [reply Formula/deno.rb]: thanks!

**Comment by carlocab** [comment]
Build failed on Linux/ARM.

**Comment by lucacasonato** [comment]
@carlocab Deno only supports ARM Linux as a "best effort support" target, and it is currently broken. Can we disable ARM Linux for now?

**Comment by cho-m** [comment]
Linux failure is probably the same V8 error as #101750 due to Homebrew's older glibc 2.23. Need to check if possible to temporarily add similar patch to formula for V8 inside `deno` until Homebrew migrates Linux image from Ubuntu 16.04 to 18.04

**Comment by danielnachun** [comment]
Can we build `rusty_v8` as a resource first and then force `deno` to use it?  Or better yet build `rusty_v8` against brewed V8?  This is a really great example of why projects should always provide a way to use the system version of dependencies rather than vendoring their own copies.  We've already solved this issue in our `v8` formula and shouldn't have to fix it again in another formula.

**Comment by aslilac** [comment]
A couple of us spent some time debugging this this morning. The build issues on macOS seem to be a mismatch between the version of LLVM (which is coming from Homebrew) and the objects being linked (which are coming from the system).

Saying `ENV["CLANG_BASE_PATH"] = "/Library/Developer/CommandLineTools/usr"` fixes the build (at least on Intel macOS)

**Comment by danielnachun** [comment]
> A couple of us spent some time debugging this this morning. The build issues on macOS seem to be a mismatch between the version of LLVM (which is coming from Homebrew) and the objects being linked (which are coming from the system).
> 
> Saying `ENV["CLANG_BASE_PATH"] = "/Library/Developer/CommandLineTools/usr"` fixes the build (at least on Intel macOS). I'll try it on my Apple Silicon machine in a bit and update with how it goes.

Ugh, that sounds like it's the same issue we've been discussing here: https://github.com/Homebrew/homebrew-core/pull/97618.  Basically we need to change the way we ship LLVM in Homebrew (at least on macOS) so that its copy of `libc++` is "hidden" from other packages and the system `libc++` from macOS is always used.  We don't have this issue on Linux because we use `libstdc++` for everything which will never be mixed up with `libc++`.  We need to ship LLVM 14 first and probably also separate out LLDB from the rest of LLVM before we can address this, but we should be close to getting both of those done.

This fix makes sense and we should use it for now if it works, but hopefully we won't need to deal with these issues in the future because they are very difficult to debug!

**Comment by cho-m** [comment]
ARM failure looks like previous V8 formula error which was worked around by disabling `lld` https://github.com/Homebrew/homebrew-core/pull/94338#issuecomment-1028480839. From other PR, it sounds like V8 issue was ARM-only.

https://github.com/Homebrew/homebrew-core/blob/bada981ad9fb0fe0013d3346add76e66a66e8670/Formula/v8.rb#L120

So, another option would be to try `ENV["GN_ARGS"] = "use_lld=false"` in `deno`.

---

Though, not sure if any side effect of issue in #96915.

Also, LLVM 14 is almost ready if we want to do a run with that before any other changes.

**Comment by danielnachun** [comment]
It's possible this a situation where `ld` can better handle the problem in https://github.com/Homebrew/homebrew-core/issues/96915 than `lld`.

I wonder if the fact `ENV["CLANG_BASE_PATH"] = "/Library/Developer/CommandLineTools/usr"` also seems to fix the problem might indicate that `v8` is trying to use the whole `clang` tool chain from LLVM.

I have a more basic question: can this (and `v8`) just use `uses_from_macos "llvm" => :build"?  Obviously we have to provide LLVM on Linux but is there something we are providing with the LLVM formula that isn't available from macOS?

**Comment by aslilac** [comment]
> So, another option would be to try `ENV["GN_ARGS"] = "use_lld=false"` in deno.

Already tried that locally, doesn't fix Apple Silicon builds unfortunately 🥲 I am curious about testing with LLVM 14 tho

**Comment by aslilac** [comment]
Side note: we should remove the `if` statement at the beginning of `install`, and the `-j1` flag from `cargo`. Those are old and unnecessary.

**Comment by cho-m** [comment]
> Already tried that locally, doesn't fix Apple Silicon builds unfortunately 🥲 I am curious about testing with LLVM 14 tho

Odd. I was able to build on Apple Silicon after only adding `ENV["GN_ARGS"] = "use_lld=false"`.

---

> I have a more basic question: can this (and `v8`) just use `uses_from_macos "llvm" => :build"`? Obviously we have to provide LLVM on Linux but is there something we are providing with the LLVM formula that isn't available from macOS?

I see that `deno` build (ignoring V8 for now) uses `lld` for Apple Silicon. Unless we override/remove that, we need `lld` available. 

https://github.com/denoland/deno/blob/v1.23.1/.cargo/config#L13-L14
```toml
[target.aarch64-apple-darwin]
rustflags = ["-C", "link-arg=-fuse-ld=lld"]
``` 

Another question is support range of V8. The original issue this PR hit was because V8 was using flags from LLVM/Clang 15. So, the farther back we go, the higher chances of unsupported options. We could try a version restriction if we know particular Apple Clang versions that are supported (e.g. `, since: :big_sur`). It looks like we used to do something like this a while back but removed it to simplify logic (#52562).

As note, I was able to build with Apple Clang/LLVM here after dropping the .cargo config and using
```
ENV["GN_ARGS"] = "use_lld=false"
ENV["CLANG_BASE_PATH"] = "/Library/Developer/CommandLineTools/usr"
```

**Comment by aslilac** [comment]
> Odd. I was able to build on Apple Silicon after only adding ENV["GN_ARGS"] = "use_lld=false".

Huh, weird

We actually just got it working on Apple Silicon with a small upstream patch to rusty_v8, fixing some gn config

**Comment by carlocab** [comment]
> Side note: we should remove the `if` statement at the beginning of `install`, and the `-j1` flag from `cargo`. Those are old and unnecessary.

Is it unnecessary? The issue I have about it is still open: https://github.com/denoland/deno/issues/9244

Also, I think it is just a good idea to avoid `lld` wherever possible. `lld` only recently became production-ready on macOS, and upstream projects that use it also routinely use a tip-of-tree build of LLVM (and hence get quick access to bug fixes).

**Comment by carlocab** [comment]
> > A couple of us spent some time debugging this this morning. The build issues on macOS seem to be a mismatch between the version of LLVM (which is coming from Homebrew) and the objects being linked (which are coming from the system).
> > Saying `ENV["CLANG_BASE_PATH"] = "/Library/Developer/CommandLineTools/usr"` fixes the build (at least on Intel macOS). I'll try it on my Apple Silicon machine in a bit and update with how it goes.
> 
> Ugh, that sounds like it's the same issue we've been discussing here: #97618. Basically we need to change the way we ship LLVM in Homebrew (at least on macOS) so that its copy of `libc++` is "hidden" from other packages and the system `libc++` from macOS is always used. 

I'm not sure it's the same, since we don't have linkage with `libc++` at all in the existing `deno` bottles.

**Comment by aslilac** [comment]
> Is it unnecessary? The issue I have about it is still open

I don't think anyone internally has run into this in a while, but I could be wrong. Hard thing about this kind of error is that's it's hard/impossible to tell if it's actually "fixed". 🤷‍♀️

**Comment by carlocab** [comment]
> I don't think anyone internally has run into this in a while

I was under the impression that no one internally was ever really that troubled by it. I'm guessing the issue has to do with, or at least is exacerbated by, virtualisation. I just remember almost always getting random build failures with `deno` until we got rid of parallelisation.

**Comment by cho-m** [comment]
Updating LLVM has fixed `lld` issue on ARM.

Still need to find a way to work around Linux issue. V8 has decided to drop `MFD_CLOEXEC` usage that is causing issue with Glibc 2.23 (https://github.com/v8/v8/commit/8fdb91cdb80ae0dd0223c0d065f724e480c5e0db) however the commit is on 10.5.x (Homebrew's v8 is Chrome's stable 10.2.x while Deno's v8 is 10.4.x). Not sure if they backport changes. Things we can try:
- Maybe see if we can use `CPPFLAGS` to pass in `-DMFD_CLOEXEC=...`
- Temporarily download copy of `rusty_v8`, locally patch copy of `v8` submodule, and then add an override to Cargo.toml via `[patch]` section

**Comment by danielnachun** [comment]
This is good news from upstream:
> The flag is causing build failures in some environments (when it is not
defined) and it is not actually needed since the behaviour of these FDs
across fork/exec isn't defined and this code is only used for testing.

This means that we don't have to worry about anything breaking on older kernels if this is undefined, and we may be able to drop the patches we have in `v8` and `qt` for this as well, once the changes have made into into the respective packages.

> - Maybe see if we can use CPPFLAGS to pass in -DMFD_CLOEXEC=...
> - Temporarily download copy of rusty_v8, locally patch copy of v8 submodule, and then add an override to Cargo.toml via [patch] section

I think we should see if the first option works.  I didn't think to try to just defining these as CPPFLAGS in `v8` and `qt`, but that would avoid the trouble of patching entirely.

**Comment by cho-m** [comment]
Tried env flags but I am guessing `gn` is designed to ignore all these.

Unless someone knows of any way to do this in `gn`, I think we'll have to either manually patch or see if `deno` is willing to float this v8 patch as backport.

**Comment by aslilac** [comment]
> see if deno is willing to float this v8 patch

You're talking about the `-D` flag yeah? I think we can do that, I'll look into it tomorrow

**Comment by danielnachun** [comment]
> > see if deno is willing to float this v8 patch
> 
> You're talking about the `-D` flag yeah? I think we can do that, I'll look into it tomorrow
> 
> (We might also be able to just upgrade to V8 10.5)

If that can be passed in through the `gn` argument then hopefully this would fix the failure.  If you're able to update to V8 10.5 that would be even better, though I have no idea how much work that is on your end.  It might just be faster though to backport https://github.com/v8/v8/commit/8fdb91cdb80ae0dd0223c0d065f724e480c5e0db as a patch into your copy of V8, as that change is trivial.

**Comment by cho-m** [comment]
Added a different workaround for Linux to apply our `v8` formula patch. The official commit is harder to inject as it is trickier to `inreplace` multiline and we can't use `patch` DSL on a .crate. It would require either manually running patch command or using Git source, which adds some overhead due to submodule fetching.

On `deno` side, it would be better to get official commit either via backport or 10.5

**Comment by aslilac** [comment]
Oh, and nice! LLVM 14 landed! I think this patch looks good for now. I'll work on fixing this on our end for 1.23.2, but that'll be either Wednesday or Thursday.

I guess ultimately it's up to you guys; if you're okay with this current fix for a couple days, or if you'd rather wait for us. (I think I'd lean towards merging as is, since 1.22.0 is pretty old at this point.)

**Comment by danielnachun** [comment]
Since all these fixes are temporary anyway, I'm personally fine with merging as-is.  This PR has been stuck because of an unusual convergence of unrelated issues with `v8`, `LLVM` and `rust` and since @cho-m did a great job documenting what needs to be reverted, I think it should be easy to remove the workarounds whenever 1.23.2 is ready.  Given that @aslilac is an upstream developer actively working with us to fix this, we don't have to worry about carrying a patch around for months/years.

**Review by danielnachun — APPROVED** [review]

**Pull request closed without merging**